Implement car

mladjo2505 opened the pull request:
https://trello.com/c/mW5PAJHu/493-lodo-car-md-car

https://trello.com/c/CTQdeGGa/274-lodo-car-department-selected-in-fb-should-be-automatically-downloaded-to-lodo

mladjo2505 committed:
Merge branch 'master' into md/car

mladjo2505 committed:
Add: Migration for Car

mladjo2505 committed:
Add: Car views

mladjo2505 committed:
Remove: Car related info from department view

mladjo2505 committed:
Add: Car select element form helpers

mladjo2505 committed:
Add: Car to vouchers(journal)

mladjo2505 committed:
Add: Car to accountplan views

mladjo2505 committed:
Add: Car to bank views
Comment out one debug print statement
Fix bank.tabsimple view

mladjo2505 committed:
Add: Car to invoicein view, also include Car in FB invoicein download

mladjo2505 committed:
Add: Car to salary views

mladjo2505 committed:
Add: Car to postmotpost report

mladjo2505 committed:
Add: Car to all other report views

mladjo2505 committed:
Revert "Remove: Car related info from department view"

This reverts commit e4cf186f8968393df755f9b9cefa2973467ef995.

mladjo2505 committed:
Whitspace trim

sls-lnluksa commented on db/changes/116_add_car_table.sql:
```diff
@@ -0,0 +1,48 @@
+-- Create new tables for Car and Cars milage
+CREATE TABLE IF NOT EXISTS companycar (
+CompanyCarID int(11) NOT NULL AUTO_INCREMENT,
```
Rename CompanyCar into Car. (Table name and id fields)

sls-lnluksa commented on db/changes/116_add_car_table.sql:
```diff
@@ -0,0 +1,48 @@
+PurchasePrice decimal(16, 2) DEFAULT '0',
+PurchaseAddCost int(11) DEFAULT '0',
+PurchaseYear smallint(6) NOT NULL DEFAULT '0',
+CurrentPrice int(11) NOT NULL DEFAULT '0',
```
Remove
+PurchaseAddCost int(11) DEFAULT '0',
+PurchaseYear smallint(6) NOT NULL DEFAULT '0',
+CurrentPrice int(11) NOT NULL DEFAULT '0',

sls-lnluksa commented on db/changes/116_add_car_table.sql:
```diff
@@ -0,0 +1,48 @@
+CompanyCarID int(11) NOT NULL,
+MilageYear smallint(6) NOT NULL DEFAULT '0',
+km0101 int(11) DEFAULT '0',
+km3112 int(11) DEFAULT '0',
```
Rename km0101 to StartMilage
Rename km3112 to EndMilage

sls-lnluksa commented:
:exclamation: 
- [x] Checkout why CarID is not saved in vouchers
- [x] Checkout why CarID is not saved/displayed in invoicein
- [x] bank.tabjournal change colspan

sls-lnluksa commented on modules/car/view/record.inc:
```diff
@@ -0,0 +1,62 @@
+  $result = $_lib['db']->db_query($query);
+}
+
+if (!$_lib['input']->getProperty('action_car_delete') && $CompanyCarID) {
```
Comment why this is here. 
If you change purchase year to newer year, implement logic around it so we delete all mileage records for previous years.

sls-lnluksa commented on modules/car/view/edit.php:
```diff
@@ -0,0 +1,176 @@
+#Do car calculations
+$car_calculations = array();
+foreach($car_milage as $milage_year => $milage) {
+  if($milage->km0101) {
```
Include this logic for StartMilage = 0

sls-lnluksa commented on modules/car/view/edit.php:
```diff
@@ -0,0 +1,176 @@
+      <tr><td class="menu">Money spent on fuel</td></tr>
+      <tr><td class="menu">Money spent per mile</td></tr>
+    </table>
+  </td>
```
@stoivo Translate this.

sls-lnluksa commented on modules/car/view/edit.php:
```diff
@@ -0,0 +1,176 @@
+      </tr>
+      <tr>
+        <? foreach($_years as $year) { ?>
+          <td><input type="text" name="companycarmilage.km0101.<? print $year; ?>" value="<? print $car_calculations[$year]['km0101']; ?>" size="15"></td>
```
align all fields to the right

sls-lnluksa commented on modules/car/view/edit.php:
```diff
@@ -0,0 +1,176 @@
+    <td><input type="text" name="companycar.PurchasePrice" value="<? print $_lib['format']->Amount($companycar->PurchasePrice) ?>" size="60"></td>
+</tr>
+<tr>
+    <td class="menu">Kj&oslash;psdato</td>
```
Kj&oslash;psdato => Salgsdato

sls-lnluksa commented on modules/car/view/list.php:
```diff
@@ -0,0 +1,65 @@
+  <tr>
+    <td class="menu">Bil</td>
+    <td class="menu">Bil navn</td>
+    <td class="menu">Aktiv</td>
```
Add CarCode, PurchaseDate, SaleDate

sls-lnluksa commented on modules/car/view/record.inc:
```diff
@@ -0,0 +1,62 @@
+elseif($_lib['input']->getProperty('action_car_new'))
+{
+
+	if($_lib['input']->getProperty('companycar_CompanyCarID') > 0) {
```
Redirect to car.list if there is no CarID.
Assign property as variable for later use.

sls-lnluksa commented on modules/car/view/record.inc:
```diff
@@ -0,0 +1,62 @@
+			$post['companycar_ValidFrom'] = "NOW()";
+			$post['companycar_Active']  = 1;
+
+			$query  = "select CompanyCarID from $db_table where CompanyCarID=" . (int) $_lib['input']->getProperty('companycar_CompanyCarID') . "";
```
Change this variable to insert query from line below.

sls-lnluksa commented on modules/car/view/record.inc:
```diff
@@ -0,0 +1,62 @@
+		else
+		{
+			$CompanyCarID = $_POST['companycar_CompanyCarID'];
+		}
```
Put `$CompanyCarID = $_POST['companycar_CompanyCarID'];` outside if block and remove else block.

sls-lnluksa commented on modules/invoicein/view/edit.php:
```diff
@@ -224,6 +225,7 @@
         </td>
         <td><? print $_lib['form3']->text(array('table'=>$db_table2, 'field'=>'ProductNumber', 'pk'=>$LineID, 'value'=>$row2->ProductNumber, 'width' => 20, 'maxlength' => 20, 'tabindex'=>$tabindex++)) ?></td>
         <td><? print $_lib['form3']->text(array('table'=>$db_table2, 'field'=>'ProductName'  , 'pk'=>$LineID, 'value'=>$row2->ProductName,   'width' => 20, 'maxlength' => 80, 'tabindex'=>$tabindex++)) ?></td>
+        <td><? if($row2->CarID) { ?><? $_lib['form2']->car_menu2(array('table' => $db_table, 'field' => 'CarID', 'value' => $row2->CarID, 'tabindex' => $tabindex++)); } ?></td>
```
If reskonto has car enabled, than show it.

sls-lnluksa commented on modules/report/view/reskontrovoucherprint.php:
```diff
@@ -193,6 +206,13 @@ function get_saldo($account, $fromperiod, $toperiod)
+            if (!$_reskontroFrom && !$_reskontroTo) {
+              $acc_start_end  = $_lib['db']->get_row(array("query" => "SELECT MIN(AccountPlanID) AS MinAccPlanID, MAX(AccountPlanID) AS MaxAccPlanID FROM accountplan WHERE AccountPlanType = '$_type'"));
+              $_reskontroFrom = $acc_start_end->MinAccPlanID;
+              $_reskontroTo   = $acc_start_end->MinAccPlanID;
```
`$_reskontroTo   = $acc_start_end->MinAccPlanID;` should be MaxAccPlanID

mladjo2505 committed:
Rework after code review

stoivo committed:
Tranlations

stoivo commented:
Test out with @hildegunn95

stoivo commented:
Remaining issues:
Car edir:
- [x] 1.new line for every 10th year
- [x] 2.add Kjøretøygruppe (Car type), dropdown,
- [x] 3.save button on left side
- [x] 4. if purchase date is "Simon" don't create YearMilage from year 0

Other
- [x] Add CarID here before Prosj http://puu.sh/l8lVW/f679604917.png

What is "Aktiver MVA"? just explain it so we know

Edit: 2. More info in next comment

stoivo commented:
This is the new fields we want:

```
Registration number         Registreringsnummer
Car brand and modell        Merke og modell
Type of car             Kjøretøygruppe (dropdown)**
Registration year           Registreringsår
Amount of seats         Antall seter
Fuel                        Drivstoff

**
Varebil(klasse 2) 
Personbil
Lastebil(tankbil)
Slepvogn
```

mladjo2505 committed:
Add new things and fix errors

Add 4 new fields for Car
Split calculations table 10 years per line
Add car to one report where it was missing

mladjo2505 committed:
Merge branch 'md/car' of https://github.com/dodoas/lodo into md/car

sls-lnluksa commented:
lodo.php?t=car.edit
Drivstoff should be select between DIESEL, BENSIN or ELEKTRISK
Slutt milage => Slutt km.stand

mladjo2505 committed:
Fix translation error
Change Fuel field to dropdown select for fuel type

hildegunn95 commented:
- [x] Registreringsår only 4 digits (registration year)
- [x] Drivstoff is not displayed after it’s saved
- [x] Typen => Type
- [x] Bil navn => Bilnavn
- [x] Format number in km.stand
- [x] Round 'kr per kilometer' to two decimals
- [x] Saldo balance, 'Not selected car'(ikke valgt) option in dropdown for cars
- [x] Remove car from reskontro ‘saldobalanse’, ‘bilagsutskrift’ and ‘åpne poster’.
- [x] Select a car on a journal and save. Than make the car inactive. Refresh the journal edit and see. Now we can not see the car, something we should be able to. And if re-activate the car, the car appears in the journal again.

mladjo2505 committed:
Fixes as requested:
Registration year only 4 digits
Change text for Type and Car name
Formatting numbers in calculation
Add diverse car, make it default(makes using reports easier), not editable
Remove car from reskontro reports
Show inaktive cars in dropdown

stoivo commented:
- [x] All accounts between 7000 and 7099 should have car enabled
- [x] On car edit view reorder fields to: Active, Registration number, ... Number of seats, Car name, ...
- [x] Trim spaces from registration number on save (both on the edges and inside)
- [x] Voucher tax line should have the same car as it's parent line (update on change)
- [x] Realign on reports list page Reskontro > Saldobalanse: Avdeling, Prosjekt, Art should all be in the same line 
- [x] On voucher edit page if the car is selected and is not active it should be shown in the dropdown (but marked as inactive) otherwise list only active cars
- [x] Implement fetching car info from fakturabank (the unique link between fakturtabank and lodo should be registration number which needs to be without spaces)
- [x] Implement basic car info fetching on fakturabank (separate task)
- [x] Implement all these fields to fakturabank (create a separate task)

mladjo2505 committed:
Fix requested things remaining on car
